### DISCUSSION TITLE
Can't get Processor.dig to run my program

### DISCUSSION THREAD

**Discussion by jnoche** [hneemann/Digital#715] [Q&A] [answered]
I apologize in advance if this is a simple matter, but I am unable to have a program run correctly in the circuit called Processor.dig.

I use ASM 3 v0.6.1 (Windows 10, JRE 15.0.1) to create the program below and save it as temp.hex.

```
1   | 0000: 1401                LDIs  R0,1             ; 0x1
2   | 0001: 1412                LDIs  R1,2             ; 0x2
3   | 0002: 0401                ADD   R0,R1
4   | 0003: 7c00                BRK   
5   | 0004: 6ffd                JMPs  2                ; 0x2
```

I run Digital v0.27 (Windows 10, JRE 15.0.1) and open the Processor.dig file in the examples\processor folder.  I edit the ROM so that it contains temp.hex.

<img width="596" alt="ROM" src="https://user-images.githubusercontent.com/63948655/115377451-36535f80-a202-11eb-899b-43d31d1420f7.png">

I edit the clock signal so that its frequency is 1 Hz.  I edit the circuit specific settings to show the measurement graph at simulation start.  I start the simulation and I can see the wire at the ROM output show the correct temp.hex values as the clock ticks.  I then stop the simulation after a while.  The measurement graph is shown below.

<img width="279" alt="measure" src="https://user-images.githubusercontent.com/63948655/115377416-2dfb2480-a202-11eb-9028-84c5ce1941f4.png">

1.  The program counter just counts from 0 to 7 and onward.  It doesn't stop (which I thought would happen because of the BRK instruction in line 4) or go back to address 2 (which I thought would happen because of the JMPs 2 instruction in line 5).

2.  Registers R0 and R1 are unchanged, which is not expected because of the LDIs and ADD instructions in lines 1 to 3.

Could you tell me where I'm going wrong?

**Comment by hneemann** [comment, marked as answer]
This is my fault! 
I changed the processor from 6-bit opcodes to 7-bit opcodes, which made it a bit simpler. Also, this allows me to add some commands that were missing like compare with carry.
And while the latest release of the simulator/processor already contains these changes, the latest release of the assembler does not.
I will create a new release. Until then, please use this file:

[asm3.zip](https://github.com/hneemann/Digital/files/6343930/asm3.zip)

**Reply by jnoche** [reply]
Thank you for the quick response.  It's working fine, now.  (And I must also apologize that I previously misunderstood the operation of the BRK instruction.  I expected it to work in Digital, but I just discovered that it is intended for Assembler.)

**Reply by hneemann** [reply]
The BRK instruction is intended to function similarly to the BRK instruction in the x86 architecture, which is a one-byte interrupt instruction that calls an ISR implemented as part of a debugger. Only that here the debugger does not run on the same processor as the code to be debugged, but instead in the assembler. This leads to the fact that the BRK command does nothing if the program was not started from the assembler.

**Comment by hneemann** [comment]
I have created a new Assembler release.
It is not necessary to save the hex file in the assembler and reload it in the processor.
If the simulator is running and the processor is loaded it is possible to click "run" or "debug" in the assembler. 
The code is transfered to the simulator and is running there automatically. See [here](https://youtu.be/K68LnY4BOEM).

**Reply by jnoche** [reply]
Thanks!
